app/models/targeting.rb

```diff
@@ -0,0 +1,41 @@
+class Targeting < ActiveRecord::Base
+
+  STATIC_TYPE = 'static_query'
+  DYNAMIC_TYPE = 'dynamic_bookmark'
+  TYPES = {STATIC_TYPE => N_('Static Query'), DYNAMIC_TYPE => N_('Dynamic Bookmark')}
+
+  belongs_to :user
+  belongs_to :bookmark
+
+  has_many :targeting_hosts, :dependent => :destroy
+  has_many :hosts, :through => :targeting_hosts
+  has_many :template_invocations, :dependent => :destroy
+
+  validates :targeting_type, :presence => true, :inclusion => Targeting::TYPES.keys
+  validates :bookmark, :presence => true, :if => "dynamic?"
+
+  before_create :assign_search_query, :if => :static?
```

Comment:
Yeah, my suggestion was A) and C) I think but I agree that we need a way to change targets of multiple invocations on one place. I remember I raised this few months ago and you persuaded me with same arguments :-) Is it too complex and confusing to add third targeting type for C)? UI will have to have inline help anyway. Also I'm fine with answer it's easy to add later (which I think is true) :-)